x-pack/plugins/alerting/server/rules_client/rules_client.ts

```diff
@@ -1258,6 +1350,284 @@ export class RulesClient {
+        { apiKeys: apiKeysToInvalidate },
+        this.logger,
+        this.unsecuredSavedObjectsClient
+      );
+    }
+
+    const updatedRules = results.map(({ id, attributes, references }) => {
+      return this.getAlertFromRaw<Params>(
+        id,
+        attributes.alertTypeId as string,
+        attributes as RawRule,
+        references,
+        false
+      );
+    });
+
+    return { rules: updatedRules, errors, total };
+  }
+
+  private async bulkEditOcc<Params extends RuleTypeParams>({
+    filter,
+    operations,
+    paramsModifier,
+  }: {
+    filter: KueryNode | null;
+    operations: BulkEditOptions<Params>['operations'];
+    paramsModifier: BulkEditOptions<Params>['paramsModifier'];
+  }): Promise<{
+    apiKeysToInvalidate: string[];
+    rules: Array<SavedObjectsBulkUpdateObject<RawRule>>;
+    resultSavedObjects: Array<SavedObjectsUpdateResponse<RawRule>>;
+    errors: BulkEditError[];
+  }> {
+    const rulesFinder =
+      await this.encryptedSavedObjectsClient.createPointInTimeFinderDecryptedAsInternalUser<RawRule>(
+        {
+          filter,
+          type: 'alert',
+          perPage: 100,
+          ...(this.namespace ? { namespaces: [this.namespace] } : undefined),
+        }
+      );
+
+    const rules: Array<SavedObjectsBulkUpdateObject<RawRule>> = [];
+    const errors: BulkEditError[] = [];
+    const apiKeysToInvalidate: string[] = [];
+    const username = await this.getUserName();
+
+    for await (const response of rulesFinder.find()) {
+      await pMap(
+        response.saved_objects,
+        async (rule) => {
+          try {
+            if (rule.attributes.apiKey) {
+              apiKeysToInvalidate.push(rule.attributes.apiKey);
+            }
+
+            const ruleType = this.ruleTypeRegistry.get(rule.attributes.alertTypeId);
+
+            let attributes = cloneDeep(rule.attributes);
+            let ruleActions = {
+              actions: this.injectReferencesIntoActions(
+                rule.id,
+                rule.attributes.actions,
+                rule.references || []
+              ),
+            };
+            for (const operation of operations) {
+              switch (operation.field) {
+                case 'actions':
+                  await this.validateActions(ruleType, operation.value);
+                  ruleActions = this.applyBulkEditOperation(operation, ruleActions);
+                  break;
+                default:
+                  attributes = this.applyBulkEditOperation(operation, attributes);
+              }
+            }
+
+            const ruleParams = paramsModifier
+              ? await paramsModifier(attributes.params as Params)
+              : attributes.params;
+
+            // validate rule params
+            const validatedAlertTypeParams = validateRuleTypeParams(
+              ruleParams,
+              ruleType.validate?.params
+            );
+            const validatedMutatedAlertTypeParams = validateMutatedRuleTypeParams(
```

Comment:
Why do we have a separate validation function for updated rule type params? Does this leave it open for params that are invalid but valid according to this separate "mutated param validation" to get saved?

Reply:
Here is a link to technical design decision made about it https://github.com/elastic/kibana/issues/124715#issuecomment-1066092827 (second paragraph)

> First of all, we think it will be good idea to add a new property in [AlertTypeParamsValidator](https://github.com/elastic/kibana/blob/main/x-pack/plugins/alerting/server/types.ts#L124-L126) to allow rule type to manage their own validation around their params. We think validate is more here to validate the schema of the params object but validateMutatedParams will be there to do more logical validation for the need of the rule type creator.